### PR TITLE
Add `test_without_assertions` lint

### LINKED ISSUE #17488: Empty Test Lint
### What it does

This lint should suggest removing empty test functions.

### Advantage

By applying this lint it would make source code cleaner / shorted by removing tests that do nothing. 

### Drawbacks

_No response_

### Example

```rust
#[test]   
fn my_empty_test() {

}
```

Could be removed

### Comparison with existing lints

To my knowledge this lint doesn't exist yet.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by willwang-io**
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust-clippy/pull/17626)*

Fixes rust-lang/rust-clippy#17488

Tests without assertions or other potential failure sites can pass without verifying behavior. This adds `test_without_assertions`, a suspicious lint that detects these tests, including empty test bodies.

changelog: [`test_without_assertions`]: add a suspicious lint for tests without assertions or other potential failure sites

- [x] Followed lint naming conventions
- [x] Added passing UI tests, including the `.stderr` file
- [x] `cargo test` passes locally
- [x] Executed `cargo dev update_lints`
- [x] Added lint documentation
- [x] Ran `cargo dev fmt`

### CODE REVIEW

File: clippy_lints/src/empty_test.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    /// ```no_run
+    /// #[test]
+    /// fn empty_test() {}
+    /// ```
+    ///
+    /// Remove the test if it is no longer needed, or add assertions that exercise the behavior it
+    /// is intended to test.
+    #[clippy::version = "1.100.0"]
+    pub EMPTY_TEST,
+    restriction,
```
Comment (CommanderStorm) [inline clippy_lints/src/empty_test.rs]: not suspicious? 🤔

Comment (CommanderStorm) [inline clippy_lints/src/empty_test.rs]: how about moving it to susicious?
I think it is rather suspicious to have such a test 😉

```suggestion
    /// ### What it does
    /// Checks for test functions with an empty body.
    ///
    /// ### Why is this bad?
    /// Empty tests pass automatically without asserting any behavior.
    /// They often represent unfinished work or forgotten stubs, giving a false sense of completeness without adding any coverage of the code.
    ///
    /// ### Example
    /// ```no_run
    /// #[test]
    /// fn ultimate_answer() {}
    /// ```
    ///
    /// Use instead:
    /// ```no_run
    /// #[test]
    /// fn ultimate_answer() {
    ///     assert_eq!(deep_thought(), 42);
    /// }
    /// ```
    #[clippy::version = "1.100.0"]
    pub EMPTY_TEST,
    suspicious,
```

File: clippy_lints/src/empty_test.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        cx: &LateContext<'_>,
+        kind: FnKind<'_>,
+        _: &FnDecl<'_>,
+        body: &Body<'_>,
+        span: Span,
+        fn_def_id: LocalDefId,
+    ) {
+        if matches!(kind, FnKind::ItemFn(..))
+            && is_test_function(cx.tcx, fn_def_id)
+            && matches!(body.value.kind, ExprKind::Block(block, _) if block.stmts.is_empty() && block.expr.is_none())
```
Comment (CommanderStorm) [inline clippy_lints/src/empty_test.rs]: lets flip them because the matches are dirt cheap and `is_test_function` looks like it does some work (binary search in one case) internally to know it is a test fn.

```suggestion
        if let ExprKind::Block(block, _) = body.value.kind
            && block.stmts.is_empty()
            && block.expr.is_none())
            && matches!(kind, FnKind::ItemFn(..))
            && is_test_function(cx.tcx, fn_def_id)
```

File: tests/ui/empty_test.stderr [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+   = note: `-D clippy::empty-test` implied by `-D warnings`
+   = help: to override `-D warnings` add `#[allow(clippy::empty_test)]`
+
+error: empty test function
+  --> tests/ui/empty_test.rs:8:1
+   |
+LL | / fn empty_with_comment() {
+LL | |     // This is still an empty body.
+LL | | }
+   | |_^
```
Comment (CommanderStorm) [inline tests/ui/empty_test.stderr]: it would be very nice to get the `#[test]` attr into this span, but that sounds like some work, so this is also fine likely.

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
Community review

From the issue:

> This can be a more general lint for tests which can never fail or always fail. Empty tests aren't particularly unique other than being the degenerate case.

I think you missed this part.
the lint I guess can start out with only linting for the degenerate case, but it should have the ambition to also lint for the provably true/false cases..

🤔
unless..
Actually I think type tests (where you ensure that for example a types size always is below a cache line for perf) cannot happen.
But that sounds a bit like an edge case
CC @Jarcho if that would be something where a "false positive" (is it one?) is good or bad?

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17626/cec41bf1cca97d7ea9beee0cf75a6268598c2dc9..d672949644e31739e8aab80b96d452b8d06b04b4)*

File: tests/ui/empty_test.stderr (hunk 3 above)
Comment (willwang-io) [reply tests/ui/empty_test.stderr]: Thanks, I agree that including `#[test]` in the span would make the diagnostic clearer. I looked into it, but it seems that `#[test]` has already been expanded by the time this late lint runs. I left the current function span as is for now. Please let me know if there is a simpler approach I have missed.

**Comment by CommanderStorm** [comment]
and the requested scope widening from the issue?

@rustbot author

**Comment by willwang-io** [comment]
@CommanderStorm I checked the existing `assertions_on_constants` lint, and it already covers things like `assert!(true)` and `assert!(false)`, and several constant-valued assertion cases. 

I am not yet sure which additional always-passing or always-falling test cases this lint should cover without overlapping with it, or flagging useful type/layout regression tests. Could you point me to the intended cases for this PR? I can then extend the lint accordingly.

**Comment by CommanderStorm** [comment]
I mean a testcase without assertions or other without panic sites would be my idea. 
Do note that there are tests that have should panic active, I don't think we should lint on those, except for empty and if non-empty.

**Comment by willwang-io** [comment]
@CommanderStorm Thanks, that makes sense now. I’ll widen this to cover tests with no assertions or other visible panic sites, while treating ordinary calls as possible panic sites.

For `#[should_panic]` tests, I’ll skip the broader check since the panic may come from a call. Should an empty `#[should_panic]` test like this be skipped too?

```rs
#[test]
#[should_panic]
fn empty_should_panic() {}
```

It already fails because it returns normally instead of panicking.

**Comment by CommanderStorm** [comment]
I think lining on this since empty is fair too. It is suspicious in any case

**Comment by willwang-io** [comment]
@CommanderStorm It looks like this newly added lint is causing some existing UI and ui-toml tests to fail. I’m not sure that adding`#![allow(clippy::test_without_assertions)]` to unrelated fixtures is the right approach.

Is that expected for a new lint in this category, or is there a different approach you would prefer?

File: clippy_lints/src/test_without_assertions.rs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                is_panic(cx, macro_call.def_id)
+                    || matches!(
+                        cx.tcx.get_diagnostic_name(macro_call.def_id),
+                        Some(
+                            sym::assert_macro
+                                | sym::assert_eq_macro
+                                | sym::assert_ne_macro
+                                | sym::debug_assert_macro
+                                | sym::debug_assert_eq_macro
+                                | sym::debug_assert_ne_macro
```
Comment (CommanderStorm) [inline clippy_lints/src/test_without_assertions.rs]: We nowerdays also have assert_matches and it's debug variant. Could you add a case and testcase for this too?

File: clippy_lints/src/test_without_assertions.rs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                );
+            }
+        }
+    }
+}
+
+fn has_potential_failure_site(cx: &LateContext<'_>, body: &Body<'_>) -> bool {
+    for_each_expr_without_closures(body.value, |expr| {
+        let is_failure_site = matches!(expr.kind, ExprKind::Call(..) | ExprKind::MethodCall(..))
+            || root_macro_call_first_node(cx, expr).is_some_and(|macro_call| {
```
Comment (CommanderStorm) [inline clippy_lints/src/test_without_assertions.rs]: We need to explicitly state this false negative of not descending into the called functions in the known problems sections or fix it

File: tests/ui/test_without_assertions.rs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+fn ordinary_empty_function() {}
+
+#[test]
+fn no_failure_site() {
+    let _ = 1 + 1;
+}
+//~^^^ test_without_assertions
+
+#[allow(clippy::unused_unit)]
```
Comment (CommanderStorm) [inline tests/ui/test_without_assertions.rs]: Please use expect instead of allow, since it is a tighter fit

File: tests/ui/test_without_assertions.rs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+#[test]
+#[should_panic]
+fn should_panic_without_visible_panic() {
+    let _ = 1 + 1;
+}
+
+#[test]
+fn calls_helper() {
+    helper();
+}
```
Comment (CommanderStorm) [inline tests/ui/test_without_assertions.rs]: Please annotate that this is a false positive and use black_box for another test where recursing into FN Def's is not possible and one test.

File: tests/ui/test_without_assertions.rs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+#[test]
+fn debug_assertion() {
+    let val = 1 + 1;
+    debug_assert!(val == 2);
+}
+
+fn helper() {}
+
+fn main() {}
```
Comment (CommanderStorm) [inline tests/ui/test_without_assertions.rs]: Our usual strategy (a bit inconsistent) in this case is to move everything into an `mod main`.
This way, all followups can be nicely placed into issue_* modules ^^

Could you do this?

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
Nice progress. 
I have some minor comments below, but nothing that is hard to fix

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17626/464e28509d0fdb5e409a5163a74a6b4cbdf0c26d..0c06278be363b69ee5fd76b13ba835c3b1944453)*

**Comment by CommanderStorm** [comment]
I flew over the changes in assertions. 
Nothing jumped out at me. 
Make sure to use expect(..) instead of assert though ^^
Adding expect(..) to lines in the code that are truely violating this is a an indicator that the lint is working and being useful imo!

Also run an lintcheck run to see if I missed a case that should actually lint in your opinion upon second view. 
I am in a plane for the next few hours, so can't do this, sorry 😔

File: tests/ui/test_without_assertions.rs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    fn calls_helper() {
+        // False negative: calls are treated as potential failure sites without inspecting their bodies.
+        helper();
+    }
+
+    #[test]
+    fn calls_black_box() {
+        // False negative: external function bodies cannot be inspected.
+        black_box(());
+    }
```
Comment (CommanderStorm) [inline tests/ui/test_without_assertions.rs]: Is it a false negative though? We should not be able to know anything about what is past the black box or?
Why do you think this is a false negative (you might be right, I might be wrong)?
If you are right on this and this acts as just another fun to the compiler then we should just remove the above... 

If we cannot look past here it is impossible to reason about this case though => we should not lint because of this.

File: clippy_lints/src/test_without_assertions.rs [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    })
+    .is_some()
+}
+
+fn is_assert_matches(cx: &LateContext<'_>, def_id: DefId) -> bool {
+    cx.tcx.crate_name(def_id.krate) == sym::core
+        && matches!(
+            cx.tcx.item_name(def_id).as_str(),
+            "assert_matches" | "debug_assert_matches"
+        )
```
Comment (CommanderStorm) [inline clippy_lints/src/test_without_assertions.rs]: Why? Why not like above?

**Review by CommanderStorm — CHANGES_REQUESTED** [review]
I have two comments that popped up because of how this is implemented 

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17626/464e28509d0fdb5e409a5163a74a6b4cbdf0c26d..6bd0e3e2f13e257bff6edd2d53e27b51f9386a97)*

File: tests/ui/test_without_assertions.rs (hunk 9 above)
Comment (willwang-io) [reply tests/ui/test_without_assertions.rs]: Ah, my bad, I mixed up the black_box case with the local helper case. Since we cannot inspect past black_box, not linting there is the sensible conservative choice. I’ll remove that test/comment.

The local-helper case is the one I meant as a false negative, since we currently do not inspect even available helper bodies.

File: clippy_lints/src/test_without_assertions.rs (hunk 10 above)
Comment (willwang-io) [reply clippy_lints/src/test_without_assertions.rs]: Because `sym::assert_matches` and `sym::debug_assert_matches` do not seem to exist...

Comment (CommanderStorm) [reply clippy_lints/src/test_without_assertions.rs]: I guess yea, and sym is auto-generated. but why is `assert_matches` not in there despite the others being in there? Is this an indicator of an bug?

Comment (willwang-io) [reply clippy_lints/src/test_without_assertions.rs]: I looked into core and noticed that `assert_matches!` and `debug_assert_matches!` do not appear to have `#[rustc_diagnostic_item]` annotations, unlike the other assertion macros. I also see that `sym` entries come from a separate `rustc_span` symbol list, so I cannot tell whether the two omissions are related. Could this be an oversight?

Comment (CommanderStorm) [reply clippy_lints/src/test_without_assertions.rs]: Good freacking question, I have no idea. Maybe anoteer community reviewer can explain this. Otherwise there is still zulip.

File: clippy_lints/src/test_without_assertions.rs [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+                );
+            }
+        }
+    }
+}
+
+fn has_potential_failure_site(cx: &LateContext<'_>, body: &Body<'_>) -> bool {
+    for_each_expr_without_closures(body.value, |expr| {
+        let is_failure_site = matches!(expr.kind, ExprKind::Call(..) | ExprKind::MethodCall(..))
+            || root_macro_call_first_node(cx, expr).is_some_and(|macro_call| {
```
Comment (Jarcho) [inline clippy_lints/src/test_without_assertions.rs]: This check doesn't actually do anything. All these macros contain a function call in the expansion.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17626/464e28509d0fdb5e409a5163a74a6b4cbdf0c26d..87cc27032a5873d8b9fe221f55a84a4da1b34946)*

File: clippy_lints/src/test_without_assertions.rs [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+                                | sym::debug_assert_eq_macro
+                                | sym::debug_assert_ne_macro
+                        )
+                    )
+                    || is_assert_matches(cx, macro_call.def_id)
+            });
+        if is_failure_site {
+            ControlFlow::Break(())
+        } else {
+            ControlFlow::Continue(Descend::Yes)
```
Comment (Jarcho) [inline clippy_lints/src/test_without_assertions.rs]: `Descend` isn't needed here. You can just use `()`.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17626/464e28509d0fdb5e409a5163a74a6b4cbdf0c26d..87cc27032a5873d8b9fe221f55a84a4da1b34946)*

File: clippy_lints/src/test_without_assertions.rs (hunk 10 above)
Comment (willwang-io) [reply clippy_lints/src/test_without_assertions.rs]: Asked in Zulip: https://rust-lang.zulipchat.com/#narrow/channel/182449-t-compiler.2Fhelp/topic/Diagnostic.20items.20for.20.60assert_matches.60.20macros/with/619924566

File: clippy_lints/src/test_without_assertions.rs [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        &mut self,
+        cx: &LateContext<'_>,
+        kind: FnKind<'_>,
+        _: &FnDecl<'_>,
+        body: &Body<'_>,
+        span: Span,
+        fn_def_id: LocalDefId,
+    ) {
+        if let ExprKind::Block(block, _) = body.value.kind
+            && matches!(kind, FnKind::ItemFn(..))
```
Comment (Jarcho) [inline clippy_lints/src/test_without_assertions.rs]: This should be checked first.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17626/464e28509d0fdb5e409a5163a74a6b4cbdf0c26d..87cc27032a5873d8b9fe221f55a84a4da1b34946)*

**Review by Jarcho — CHANGES_REQUESTED** [review]

File: tests/ui/single_call_fn.rs [hunk 14]
```diff
@@ -111,3 +111,5 @@ fn m<T: Trait>(v: T) {
     v.foo();
     unsafe { rand() };
 }
+
+//@ compile-flags: -Aclippy::test_without_assertions
```
Comment (CommanderStorm) [inline tests/ui/single_call_fn.rs]: why???

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17626/464e28509d0fdb5e409a5163a74a6b4cbdf0c26d..2a24e75f78e1b2e625dcc3562d211bef088573db)*

Comment (willwang-io) [reply tests/ui/single_call_fn.rs]: Because this new lint caused some existing tests for unrelated lints to fail, making CI unhappy.. I tried to disable it only in those fixtures. Would there be a better way to handle this?

Comment (CommanderStorm) [reply tests/ui/single_call_fn.rs]: please add #[expect(..)] where it is failing, or at the top of the file if it is failing for most items ("where it makes sense")

File: tests/ui/dbg_macro/dbg_macro.rs [hunk 15]
```diff
@@ -89,6 +89,7 @@ mod issue7274 {
     });
 }
 
+#[allow(clippy::test_without_assertions)]
```
Comment (CommanderStorm) [inline tests/ui/dbg_macro/dbg_macro.rs]:
```suggestion
#[expect(clippy::test_without_assertions)]
```

**Review by CommanderStorm — APPROVED** [review]
Community review:
Lgtm. One typo fix ^^
Make sure to squash your changes into one commit. 

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17626/a73575ac42dfe5593471b0912201081df844d6ed..092ec9df1732ab3eeef99a32feccfe7404448ddf)*

File: tests/ui/dbg_macro/dbg_macro.rs (hunk 15 above)
Comment (willwang-io) [reply tests/ui/dbg_macro/dbg_macro.rs]: It looks like `#[expect]` is unfulfilled in the original `dbg!` fixture, since the lint only fires after rustfix. Would `#[allow]` make sense for this case?

Comment (CommanderStorm) [reply tests/ui/dbg_macro/dbg_macro.rs]: ah okay, yes it makes. Consider adding an reason="" please 😉

**Review by DanielEScherzer — APPROVED** [review]
Community review: looks good to me
Looked into why there wasn't a lintcheck report - its because thee were no new lint hits, which makes sense because code doesn't normally include no-op tests

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/17626/a73575ac42dfe5593471b0912201081df844d6ed..d9a725d4d64937bcd3ca5ec395ddd18640ea5a51)*
